### DISCUSSION TITLE
How do I root this custom img with magisk

### DISCUSSION THREAD

**Discussion by ParkGyuOh** [Archfx/duo-de#33] [Q&A] [answered]
I was wondering how do u root this with magisk or any other way? Perhaps an rooted custom rom?

**Comment by ParkGyuOh** [comment]
Or maybe just release the int_boot.img so we could use magisk to root?

**Reply by Archfx** [reply, marked as answer]
I used to have magisk with this ROM a long time ago, since I don't remember the exact steps I rely on @Ethanol10's answer. 

>You can use payload-dumper-go to get the boot.img out from a stock img. once you do that you can transfer the boot img to your device, patch it, then flash it.

Follow the instructions from [here](https://topjohnwu.github.io/Magisk/install.html). You should be good to go. 

PS: If you choose to do this, carefull when using OTA, the device will brick if you don't follow the steps [here](https://topjohnwu.github.io/Magisk/ota.html)! 

Good luck!

**Reply by ParkGyuOh** [reply]
That's won't working with Android 12 recovery image it has to be 24

**Reply by ParkGyuOh** [reply]
Please help to root and release with rooted rom thx

**Reply by Ethanol10** [reply]
@ParkGyuOh You can use payload-dumper-go to get the boot.img out from a stock img. once you do that you can transfer the boot img to your device, patch it, then flash it.

**Reply by Zora-Z0** [reply]
It seems like I was able to get my DUO 2 rooted below are the steps I followed.

**_PLEASE NOTE THAT THIS WILL MOST LIKELY DELETE EVERYTHING ON YOUR PHONE SO BACK UP ANY DATA YOU DO NOT WANT TO LOSE. ROOTING ALSO VOIDS YOUR WARRANTY AND YOU MIGHT BRICK YOUR DEVICE OR DAMAGE IT BEYOND REPAIR. IF YOU FOLLOW THESE STEPS YOU ARE DOING THEM AT YOUR OWN RISK AND I WILL NOT BE HELD RESPONSIBLE FOR ANY ISSUES, PROBLEMS, OR DAMAGES THAT MAY OCCUR. THIS IS MAINLY FOR MY OWN REFFRENCE AND EDUCATIONAL PURPOSES. I HAVE ONLY TRIED THIS FOR THE DUO 2 AND DO NOT KNOW IF IT WILL WORK FOR THE DUO 1. I AM NOT AN EXPERT OR KNOWLEDGEABLE  ON THIS. I ONLY FOLLOWED GUIDES ONLINE AND PIECED THINGS TOGETHER. SO I WILL NOT BE ABLE TO HELP YOU RESOLVE ANY ISSUES OR QUESTIONS YOU MIGHT HAVE. YOU WILL MOST LIKELY HAVE TO ASK SOMEONE ELSE OR SEARCH ONLINE FOR A SOLUTION._**

**_AGAIN IF YOU FOLLOW ANY OF THESE STEPS YOU ARE DOING SO AT YOUR OWN RISK._**

First I factory reset DUO-DE (not sure if its necessary). Then I downloaded the stock firmware from

https://support.microsoft.com/en-us/surface-recovery-image

(I like to rename it to surface-duo-2-recovery-image.zip)

If you don't have it you need to download android platform tools from

https://developer.android.com/tools/releases/platform-tools

You need platform tools in order to use adb commands in terminal/cmd. Then I extract the zip file and typed

_cmd_

inside the address bar of file explorer in which should open up the command prompt in that folder and then I ran this command

_adb reboot fastboot_

I followed the steps inside the discussion on how to revert back to stock but I originally had some issues like others on the DUO 2. Luckily someone figured out how to do it and here are the commands I had to use **_(REMEMBER YOU WILL LOSE ALL DATA)_**

_adb reboot fastboot_
_fastboot delete-logical-partition system_ext_
_fastboot delete-logical-partition system_ext_a_
_fastboot delete-logical-partition system_ext_b_
_fastboot delete-logical-partition product_
_fastboot delete-logical-partition product_a_
_fastboot delete-logical-partition product_b_
_fastboot delete-logical-partition system_b_
_fastboot delete-logical-partition system_a_
_fastboot reboot recovery_

Then I selected "apply update from adb" from the phone menu (using the volume keys to move the selection and using the power key to select) and typed this command but did not run it just yet

_adb sideload_ (make sure there is a space after sideload)

then dragged and dropped the stock firmware zip file into the command prompt. Once you drag and drop the zip file the command should look like

_adb sideload /path/to/file.zip_

Once it finishes installing I selected "Reboot system now" from the phone menu. It will then say that the system might be corrupted.
I then select "Factory data reset" from the phone menu **_(AGAIN YOUR ARE DELETING EVERYTHING FROM THE PHONE)_**. Most likely it will not want to boot so you will have to select recovery mode from the on screen phone prompt. Once you see the android figure press and hold the power button. Then press and release the volume up button while still holding down the power button. You should now be back in recovery mode. Then reselect "apply update from adb" You can then rerun theadb sideload command and drag and drop the stock firmware again which it should look like the previous command, for example

_adb sideload /path/to/file.zip_

(I had to install the stock firmware twice since the first time didn't work, so you might have to install it multiple times.)

This is Microsoft's guide for flashing stock firmware but it didn't work for me

https://support.microsoft.com/en-us/surface/recover-surface-duo-if-it-won-t-start-1e6fb3cf-4e24-92a8-bc05-4984274bc6a4

Next these are the tools I needed to download for rooting

Stock image of your device from Microsoft (Again you will need to have the stock image)

Android platform tools (Again you will need Android Platform Tools for ADB commands)

payload-dumper-go (payload-dumper-go is to extract the boot.img from stock firmware)

https://github.com/ssut/payload-dumper-go

7Zip (I installed 7Zip through winget to extract compressed folders/files)

F-Droid (F-droid is an alternative android app store which I downloaded and installed it to my phone to download obtainium)

https://f-droid.org

Obatinium (Obtainium lest you install apps from GitHub)
-I installed this through F-droid you just need to search for it

Magisk (magisk which is an android app used to root your phone)
(I added the following GitHub to obtainium to install it)

https://github.com/topjohnwu/Magisk/releases

Once I downloaded everything and installed everything I made sure I enabled developer mode and also made sure OEM unlocking was enable as well as USB debugging.

Now I was ready to root the phone. The guide originally provided didn't really work for me so I actually followed a YouTube guide by Tech Jarves to root the stock firmware

https://www.youtube.com/watch?v=gTd3CUT-6EE

I did some stuff differently and if the video link doesn't work here are the steps I followed. First I extracted the
payload-dumper-go_for_windows .tar.gz and inside file browser address bar I typed in

_cmd_

to open up the terminal/command prompt. Then typed this but did not run it

_payload-dumper-go_ (makes sure to leave a space after go)

Then I dragged and dropped the stock firmware zip file into the terminal and ran the command so it should look like

_payload-dumper-go /path/to/stock_image.zip_

after the command finishes running there should be a new folder called "extracted_random_numbers" Inside of the payload dumper folder you should be able to find the boot.img file inside that folder. Copy that file over to your phone. I like to put it inside the phones download folder. Then you need to open up magisk and click on install. Then tap on "select and patch a file". Then locate the boot.img and once selected it should start patching the file. Once it finishes it will then tell you where the file is located, which most likely will be inside the downloads folder, and you need to transfer it to your pc. It should be called something like

magisk_patched-random_numbers.img

Then you need to use platform tools again by typing

_cmd_

inside the address bar where the folder is located. This should open up terminal/command prompt and you need to run the following commands

_adb reboot fastboot_

_fastboot flash boot /path/to/magisk_patched-random_numbers.img_

(again you can just type in fastboot flash boot and drag and drop the magisk patched image)

Once it finishes you can reboot with

_fastboot reboot_

Once your phone reboots you may need to go through the setup wizard again. Then you can open up magisk and it should say installed "version your downloaded version" for example at this time mine says

Magisk
Installed 29.0 (29000)
Zygisk no
Ramdisk yes

This should indicate that your phone is now rooted.  Its also recommended to turn off automatic OTA updates but I didn't not find any in the DUO-DE settings so I just simply turn off the system ones, which probably doesn't matter since we cant update the stock firmware. If you are gonna update using DUO-DE OTA updates its also recommended to follow the steps linked below before running the update.

**_PLEASE NOTE I HAVE NOT TRIED THIS YET SO I DO NOT KNOW IF IT WILL WORK OR IF YOU WILL LOSE DATA OR BRICK/DAMAGE YOU PHONE SO IF YOU FOLLOW THE STEPS YOU ARE DOING SO AT YOUR OWN RISK._**

https://topjohnwu.github.io/Magisk/ota.html

In short what you need to do is open the magisk app -> uninstall -> restore images
**_BUT DO NOT REBOOT THE DEVICE!_**
Once it restores the image you can then run the OTA update and once that finishes.
**_DO NOT PRESS THE "RESTART OR REBOOT"_**
you will need to go back into Magisk App -> Install -> Install to Inactive Slot.
After installation press the reboot button **_INSIDE OF THE MAGISK APP._**
That should be the last step but again I have not tried this myself.

The guide also mentions that if its a non A/B device you will essentially have to start from the beginning and reflash the to the stock firmware to then root it again then flash the custom DUO-DE rom.

Also if anyone knows how to move apps from internal storage to external storage please let me know. All the guides I have found online are outdated and no longer work.
